File: backend/app/api/v1/favorites.py
"""API endpoints for User Favorites management."""

from typing import Annotated
from uuid import UUID

from fastapi import APIRouter, Depends, Query
from sqlalchemy import delete, func, select
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy.orm import selectinload

from app.core.deps import get_current_user
from app.core.exceptions import ConflictError, NotFoundError
from app.database import get_session
from app.models import Entity, EntityType, User
from app.models.user_favorite import UserFavorite
from app.schemas.common import MessageResponse
from app.schemas.favorite import (
    FavoriteCheckResponse,
    FavoriteCreate,
    FavoriteEntityBrief,
    FavoriteListResponse,
    FavoriteResponse,
)

router = APIRouter()


@router.get("", response_model=FavoriteListResponse)
async def list_favorites(
    page: Annotated[int, Query(ge=1, description="Page number")] = 1,
    per_page: Annotated[int, Query(ge=1, le=100, description="Items per page")] = 20,
    entity_type_slug: Annotated[str | None, Query(description="Filter by entity type slug")] = None,
    search: Annotated[str | None, Query(description="Search in entity name")] = None,
    sort_by: Annotated[str | None, Query(description="Sort by field (created_at, entity_name)")] = None,
    sort_order: Annotated[str | None, Query(description="Sort order (asc, desc)")] = "desc",
    current_user: User = Depends(get_current_user),
    session: AsyncSession = Depends(get_session),
) -> FavoriteListResponse:
    """
    List user's favorite entities with pagination and sorting.

    Supports filtering by entity type and search in entity name.
    """
    # Determine if we need an explicit Entity join for sorting/filtering
    needs_entity_join = entity_type_slug or search or sort_by == "entity_name"

    # Base query
    query = (
        select(UserFavorite)
        .where(UserFavorite.user_id == current_user.id)
        .options(selectinload(UserFavorite.entity).selectinload(Entity.entity_type))
    )

    # Join with Entity if needed
    if needs_entity_join:
        query = query.join(Entity, UserFavorite.entity_id == Entity.id)

    # Filter by entity type
    if entity_type_slug:
        subq = select(EntityType.id).where(EntityType.slug == entity_type_slug)
        query = query.where(Entity.entity_type_id.in_(subq))

    # Search in entity name
    if search:
        # Escape SQL wildcards to prevent injection
        safe_search = search.replace('%', '\\%').replace('_', '\\_')
        query = query.where(Entity.name.ilike(f"%{safe_search}%", escape='\\'))

    # Count total
    count_query = select(func.count()).select_from(query.subquery())
    total = (await session.execute(count_query)).scalar() or 0

    # Handle sorting
    sort_desc = sort_order == "desc"
    if sort_by == "entity_name":
        if sort_desc:
            query = query.order_by(Entity.name.desc(), UserFavorite.created_at.desc())
        else:
            query = query.order_by(Entity.name.asc(), UserFavorite.created_at.desc())
    else:
        # Default sort by created_at
        if sort_desc:
            query = query.order_by(UserFavorite.created_at.desc())
        else:
            query = query.order_by(UserFavorite.created_at.asc())

    # Paginate
    query = query.offset((page - 1) * per_page).limit(per_page)
    result = await session.execute(query)
    favorites = result.scalars().all()

    items = []
    for fav in favorites:
        entity = fav.entity
        entity_type = entity.entity_type if entity else None

        items.append(
            FavoriteResponse(
                id=fav.id,
                user_id=fav.user_id,
                entity_id=fav.entity_id,
                created_at=fav.created_at,
                entity=FavoriteEntityBrief(
                    id=entity.id,
                    name=entity.name,
                    slug=entity.slug,
                    entity_type_id=entity.entity_type_id,
                    entity_type_slug=entity_type.slug if entity_type else None,
                    entity_type_name=entity_type.name if entity_type else None,
                    entity_type_icon=entity_type.icon if entity_type else None,
                    entity_type_color=entity_type.color if entity_type else None,
                    hierarchy_path=entity.hierarchy_path,
                    is_active=entity.is_active,
                ),
            )
        )

    return FavoriteListResponse(
        items=items,
        total=total,
        page=page,
        per_page=per_page,
        pages=(total + per_page - 1) // per_page if per_page > 0 else 0,
    )


@router.post("", response_model=FavoriteResponse, status_code=201)
async def add_favorite(
    data: FavoriteCreate,
    current_user: User = Depends(get_current_user),
    session: AsyncSession = Depends(get_session),
) -> FavoriteResponse:
    """
    Add an entity to favorites.

    Returns 409 Conflict if the entity is already favorited.
    """
    # Verify entity exists
    entity = await session.get(Entity, data.entity_id)
    if not entity:
        raise NotFoundError("Entity", str(data.entity_id))

    # Check if already favorited
    existing = await session.execute(
        select(UserFavorite).where(
            UserFavorite.user_id == current_user.id,
            UserFavorite.entity_id == data.entity_id,
        )
    )
    if existing.scalar():
        raise ConflictError(
            "Already favorited",
            detail="This entity is already in your favorites",
        )

    # Get entity type for response
    entity_type = await session.get(EntityType, entity.entity_type_id)

    # Create favorite
    favorite = UserFavorite(
        user_id=current_user.id,
        entity_id=data.entity_id,
    )
    session.add(favorite)
    await session.commit()
    await session.refresh(favorite)

    return FavoriteResponse(
        id=favorite.id,
        user_id=favorite.user_id,
        entity_id=favorite.entity_id,
        created_at=favorite.created_at,
        entity=FavoriteEntityBrief(
            id=entity.id,
            name=entity.name,
            slug=entity.slug,
            entity_type_id=entity.entity_type_id,
            entity_type_slug=entity_type.slug if entity_type else None,
            entity_type_name=entity_type.name if entity_type else None,
            entity_type_icon=entity_type.icon if entity_type else None,
            entity_type_color=entity_type.color if entity_type else None,
            hierarchy_path=entity.hierarchy_path,
            is_active=entity.is_active,
        ),
    )


@router.get("/check/{entity_id}", response_model=FavoriteCheckResponse)
async def check_favorite(
    entity_id: UUID,
    current_user: User = Depends(get_current_user),
    session: AsyncSession = Depends(get_session),
) -> FavoriteCheckResponse:
    """
    Check if an entity is in the user's favorites.

    Returns the favorite status and favorite ID if favorited.
    """
    result = await session.execute(
        select(UserFavorite).where(
            UserFavorite.user_id == current_user.id,
            UserFavorite.entity_id == entity_id,
        )
    )
    favorite = result.scalar()

    return FavoriteCheckResponse(
        entity_id=entity_id,
        is_favorited=favorite is not None,
        favorite_id=favorite.id if favorite else None,
    )


@router.delete("/{favorite_id}", response_model=MessageResponse)
async def remove_favorite(
    favorite_id: UUID,
    current_user: User = Depends(get_current_user),
    session: AsyncSession = Depends(get_session),
) -> MessageResponse:
    """
    Remove an entity from favorites by favorite ID.
    """
    result = await session.execute(
        select(UserFavorite).where(
            UserFavorite.id == favorite_id,
            UserFavorite.user_id == current_user.id,
        )
    )
    favorite = result.scalar()

    if not favorite:
        raise NotFoundError("Favorite", str(favorite_id))

    await session.delete(favorite)
    await session.commit()

    return MessageResponse(message="Favorite removed successfully")


@router.delete("/entity/{entity_id}", response_model=MessageResponse)
async def remove_favorite_by_entity(
    entity_id: UUID,
    current_user: User = Depends(get_current_user),
    session: AsyncSession = Depends(get_session),
) -> MessageResponse:
    """
    Remove an entity from favorites by entity ID.

    Convenience endpoint for removing a favorite without knowing the favorite ID.
    """
    result = await session.execute(
        delete(UserFavorite).where(
            UserFavorite.user_id == current_user.id,
            UserFavorite.entity_id == entity_id,
        )
    )

    if result.rowcount == 0:
        raise NotFoundError("Favorite", str(entity_id))

    await session.commit()
    return MessageResponse(message="Favorite removed successfully")

File: frontend/src/stores/favorites.ts
/**
 * Favorites Store
 *
 * Manages user's favorite entities state.
 */

import { defineStore } from 'pinia'
import { ref, computed } from 'vue'
import { favoritesApi } from '@/services/api'
import { useLogger } from '@/composables/useLogger'

const logger = useLogger('FavoritesStore')

export interface FavoriteEntity {
  id: string
  name: string
  slug: string
  entity_type_id: string
  entity_type_slug: string | null
  entity_type_name: string | null
  entity_type_icon: string | null
  entity_type_color: string | null
  hierarchy_path: string | null
  is_active: boolean
}

export interface Favorite {
  id: string
  user_id: string
  entity_id: string
  created_at: string
  entity: FavoriteEntity
}

export const useFavoritesStore = defineStore('favorites', () => {
  // State
  const favorites = ref<Favorite[]>([])
  const favoriteIds = ref<Set<string>>(new Set())
  const isLoading = ref(false)
  const total = ref(0)
  const page = ref(1)
  const perPage = ref(20)
  const error = ref<string | null>(null)

  // Computed
  const favoriteCount = computed(() => favoriteIds.value.size)

  /**
   * Check if entity is favorited (instant lookup)
   */
  function isFavorited(entityId: string): boolean {
    return favoriteIds.value.has(entityId)
  }

  /**
   * Get favorite by entity ID
   */
  function getFavoriteByEntityId(entityId: string): Favorite | undefined {
    return favorites.value.find((f) => f.entity_id === entityId)
  }

  /**
   * Load favorites list from API
   */
  async function loadFavorites(options?: {
    page?: number
    per_page?: number
    entity_type_slug?: string
    search?: string
    sort_by?: string
    sort_order?: 'asc' | 'desc'
  }): Promise<void> {
    isLoading.value = true
    error.value = null

    // Update perPage if provided
    if (options?.per_page) {
      perPage.value = options.per_page
    }

    try {
      const response = await favoritesApi.list({
        page: options?.page || page.value,
        per_page: perPage.value,
        entity_type_slug: options?.entity_type_slug,
        search: options?.search,
        sort_by: options?.sort_by,
        sort_order: options?.sort_order,
      })

      favorites.value = response.data.items
      total.value = response.data.total
      page.value = response.data.page

      // Update favoriteIds set for quick lookup
      favoriteIds.value = new Set(favorites.value.map((f) => f.entity_id))
    } catch (e) {
      logger.error('Failed to load favorites:', e)
      error.value = 'Failed to load favorites'
    } finally {
      isLoading.value = false
    }
  }

  /**
   * Add entity to favorites
   */
  async function addFavorite(entityId: string): Promise<Favorite | null> {
    try {
      const response = await favoritesApi.add(entityId)
      const favorite = response.data

      // Add to local state
      favorites.value.unshift(favorite)
      favoriteIds.value.add(entityId)
      total.value++

      return favorite
    } catch (e) {
      logger.error('Failed to add favorite:', e)
      error.value = 'Failed to add favorite'
      return null
    }
  }

  /**
   * Remove favorite by entity ID
   */
  async function removeFavorite(entityId: string): Promise<boolean> {
    try {
      await favoritesApi.removeByEntity(entityId)

      // Remove from local state
      const index = favorites.value.findIndex((f) => f.entity_id === entityId)
      if (index > -1) {
        favorites.value.splice(index, 1)
      }
      favoriteIds.value.delete(entityId)
      total.value = Math.max(0, total.value - 1)

      return true
    } catch (e) {
      logger.error('Failed to remove favorite:', e)
      error.value = 'Failed to remove favorite'
      return false
    }
  }

  /**
   * Toggle favorite status for entity
   */
  async function toggleFavorite(entityId: string): Promise<boolean> {
    if (isFavorited(entityId)) {
      const success = await removeFavorite(entityId)
      return !success // Returns false if now NOT favorited
    } else {
      const favorite = await addFavorite(entityId)
      return favorite !== null // Returns true if now favorited
    }
  }

  /**
   * Check favorite status from API (for initial load on entity detail)
   */
  async function checkFavorite(entityId: string): Promise<boolean> {
    try {
      const response = await favoritesApi.check(entityId)
      if (response.data.is_favorited) {
        favoriteIds.value.add(entityId)
      } else {
        favoriteIds.value.delete(entityId)
      }
      return response.data.is_favorited
    } catch (e) {
      logger.error('Failed to check favorite status:', e)
      return false
    }
  }

  /**
   * Clear all favorites from local state (on logout)
   */
  function clearFavorites(): void {
    favorites.value = []
    favoriteIds.value.clear()
    total.value = 0
    page.value = 1
  }

  return {
    // State
    favorites,
    favoriteIds,
    isLoading,
    total,
    page,
    perPage,
    error,

    // Computed
    favoriteCount,

    // Actions
    isFavorited,
    getFavoriteByEntityId,
    loadFavorites,
    addFavorite,
    removeFavorite,
    toggleFavorite,
    checkFavorite,
    clearFavorites,
  }
})
